File: src/data/repository/WorkoutRepository.ts
/**
 * Workout Repository - Manages workout history, routines, and programs
 * Migrated from Android Kotlin WorkoutRepository to React Native TypeScript
 */

import { EventEmitter } from 'events';
import * as WorkoutDao from '../local/daos/workoutDao';
import * as PersonalRecordDao from '../local/daos/personalRecordDao';
import {
  WorkoutSessionEntity,
  WorkoutMetricEntity,
  RoutineEntity,
  RoutineExerciseEntity,
  WeeklyProgramWithDays,
  WeeklyProgramEntity,
  ProgramDayEntity,
} from '../local/entities';
import {
  WorkoutSession,
  WorkoutMetric,
  ProgramMode,
  EchoLevel,
  EccentricLoad,
  WorkoutType,
  WorkoutMode,
  WorkoutModeConstants,
} from '../../domain/models/Models';
import { Routine, RoutineExercise } from '../../domain/models/Routine';
import { Exercise, CableConfiguration } from '../../domain/models/Exercise';

/**
 * Workout Repository interface
 */
export interface IWorkoutRepository {
  // Session operations
  saveSession(session: WorkoutSession): Promise<void>;
  saveMetrics(sessionId: string, metrics: WorkoutMetric[]): Promise<void>;
  getAllSessions(): Promise<WorkoutSession[]>;
  getRecentSessions(limit?: number): Promise<WorkoutSession[]>;
  getSession(sessionId: string): Promise<WorkoutSession | null>;
  getMetricsForSession(sessionId: string): Promise<WorkoutMetric[]>;
  deleteWorkout(sessionId: string): Promise<void>;
  deleteAllWorkouts(): Promise<void>;

  // Routine operations
  saveRoutine(routine: Routine): Promise<void>;
  updateRoutine(routine: Routine): Promise<void>;
  getAllRoutines(): Promise<Routine[]>;
  getRoutine(routineId: string): Promise<Routine | null>;
  deleteRoutine(routineId: string): Promise<void>;
  markRoutineUsed(routineId: string): Promise<void>;

  // Weekly program operations
  getAllPrograms(): Promise<WeeklyProgramWithDays[]>;
  getActiveProgram(): Promise<WeeklyProgramWithDays | null>;
  getProgramById(programId: string): Promise<WeeklyProgramWithDays | null>;
  saveProgram(programWithDays: WeeklyProgramWithDays): Promise<void>;
  deleteProgram(programId: string): Promise<void>;
  activateProgram(programId: string): Promise<void>;

  // Personal record operations
  updatePersonalRecordIfNeeded(
    exerciseId: string,
    weightPerCableKg: number,
    reps: number,
    workoutMode: string
  ): Promise<boolean>;
}

/**
 * Workout Repository implementation
 */
class WorkoutRepositoryImpl extends EventEmitter implements IWorkoutRepository {
  constructor() {
    super();
  }

  // ========== Session Operations ==========

  /**
   * Save a workout session
   */
  async saveSession(session: WorkoutSession): Promise<void> {
    try {
      const entity: WorkoutSessionEntity = {
        id: session.id || this.generateUUID(),
        timestamp: session.timestamp || Date.now(),
        mode: session.mode || 'OldSchool',
        reps: session.reps || 10,
        weightPerCableKg: session.weightPerCableKg || 0,
        progressionKg: session.progressionKg || 0,
        duration: session.duration || 0,
        totalReps: session.totalReps || 0,
        warmupReps: session.warmupReps || 0,
        workingReps: session.workingReps || 0,
        isJustLift: session.isJustLift || false,
        stopAtTop: session.stopAtTop || false,
        eccentricLoad: session.eccentricLoad || 100,
        echoLevel: session.echoLevel || 1,
        exerciseId: session.exerciseId || null,
      };

      await WorkoutDao.insertSession(entity);
      console.log(`[WorkoutRepository] Saved workout session: ${entity.id}`);
    } catch (error) {
      console.error('[WorkoutRepository] Failed to save workout session:', error);
      throw error;
    }
  }

  /**
   * Save workout metrics (batch insert for performance)
   */
  async saveMetrics(sessionId: string, metrics: WorkoutMetric[]): Promise<void> {
    try {
      const entities: WorkoutMetricEntity[] = metrics.map((metric) => ({
        sessionId,
        timestamp: metric.timestamp || Date.now(),
        loadA: metric.loadA,
        loadB: metric.loadB,
        positionA: metric.positionA,
        positionB: metric.positionB,
        ticks: metric.ticks || 0,
      }));

      await WorkoutDao.insertMetrics(entities);
      console.log(`[WorkoutRepository] Saved ${entities.length} metrics for session ${sessionId}`);
    } catch (error) {
      console.error('[WorkoutRepository] Failed to save workout metrics:', error);
      throw error;
    }
  }

  /**
   * Get all workout sessions
   */
  async getAllSessions(): Promise<WorkoutSession[]> {
    try {
      const entities = await WorkoutDao.getAllSessions();
      return entities.map(this.entityToWorkoutSession);
    } catch (error) {
      console.error('[WorkoutRepository] Failed to get all sessions:', error);
      throw error;
    }
  }

  /**
   * Get recent workout sessions
   */
  async getRecentSessions(limit: number = 10): Promise<WorkoutSession[]> {
    try {
      const entities = await WorkoutDao.getRecentSessions(limit);
      return entities.map(this.entityToWorkoutSession);
    } catch (error) {
      console.error('[WorkoutRepository] Failed to get recent sessions:', error);
      throw error;
    }
  }

  /**
   * Get a specific workout session
   */
  async getSession(sessionId: string): Promise<WorkoutSession | null> {
    try {
      const entity = await WorkoutDao.getSession(sessionId);
      return entity ? this.entityToWorkoutSession(entity) : null;
    } catch (error) {
      console.error('[WorkoutRepository] Failed to get session:', error);
      throw error;
    }
  }

  /**
   * Get metrics for a workout session
   */
  async getMetricsForSession(sessionId: string): Promise<WorkoutMetric[]> {
    try {
      const entities = await WorkoutDao.getMetricsForSession(sessionId);
      return entities.map(this.entityToWorkoutMetric);
    } catch (error) {
      console.error('[WorkoutRepository] Failed to get metrics:', error);
      throw error;
    }
  }

  /**
   * Delete a workout
   */
  async deleteWorkout(sessionId: string): Promise<void> {
    try {
      await WorkoutDao.deleteWorkout(sessionId);
      console.log(`[WorkoutRepository] Deleted workout: ${sessionId}`);
    } catch (error) {
      console.error('[WorkoutRepository] Failed to delete workout:', error);
      throw error;
    }
  }

  /**
   * Delete all workouts
   */
  async deleteAllWorkouts(): Promise<void> {
    try {
      await WorkoutDao.deleteAllWorkouts();
      console.log('[WorkoutRepository] Deleted all workouts');
    } catch (error) {
      console.error('[WorkoutRepository] Failed to delete all workouts:', error);
      throw error;
    }
  }

  // ========== Routine Operations ==========

  /**
   * Save a routine with exercises
   */
  async saveRoutine(routine: Routine): Promise<void> {
    try {
      const entity = this.routineToEntity(routine);
      const exerciseEntities = routine.exercises.map((ex) => this.routineExerciseToEntity(ex, routine.id));

      await WorkoutDao.insertRoutineWithExercises(entity, exerciseEntities);
      console.log(`[WorkoutRepository] Saved routine: ${routine.name}`);
    } catch (error) {
      console.error('[WorkoutRepository] Failed to save routine:', error);
      throw error;
    }
  }

  /**
   * Update a routine
   */
  async updateRoutine(routine: Routine): Promise<void> {
    try {
      const entity = this.routineToEntity(routine);
      const exerciseEntities = routine.exercises.map((ex) => this.routineExerciseToEntity(ex, routine.id));

      await WorkoutDao.updateRoutineWithExercises(entity, exerciseEntities);
      console.log(`[WorkoutRepository] Updated routine: ${routine.name}`);
    } catch (error) {
      console.error('[WorkoutRepository] Failed to update routine:', error);
      throw error;
    }
  }

  /**
   * Get all routines
   */
  async getAllRoutines(): Promise<Routine[]> {
    try {
      const entities = await WorkoutDao.getAllRoutines();
      const routines: Routine[] = [];

      for (const entity of entities) {
        const exercises = await WorkoutDao.getExercisesForRoutine(entity.id);
        routines.push(this.entityToRoutine(entity, exercises));
      }

      return routines;
    } catch (error) {
      console.error('[WorkoutRepository] Failed to get all routines:', error);
      throw error;
    }
  }

  /**
   * Get a specific routine
   */
  async getRoutine(routineId: string): Promise<Routine | null> {
    try {
      const entity = await WorkoutDao.getRoutineById(routineId);
      if (!entity) {
        return null;
      }

      const exercises = await WorkoutDao.getExercisesForRoutine(routineId);
      return this.entityToRoutine(entity, exercises);
    } catch (error) {
      console.error('[WorkoutRepository] Failed to get routine:', error);
      return null;
    }
  }

  /**
   * Delete a routine
   */
  async deleteRoutine(routineId: string): Promise<void> {
    try {
      await WorkoutDao.deleteRoutineComplete(routineId);
      console.log(`[WorkoutRepository] Deleted routine: ${routineId}`);
    } catch (error) {
      console.error('[WorkoutRepository] Failed to delete routine:', error);
      throw error;
    }
  }

  /**
   * Mark routine as used (updates lastUsed and increments useCount)
   */
  async markRoutineUsed(routineId: string): Promise<void> {
    try {
      await WorkoutDao.markRoutineUsed(routineId);
      console.log(`[WorkoutRepository] Marked routine used: ${routineId}`);
    } catch (error) {
      console.error('[WorkoutRepository] Failed to mark routine used:', error);
      throw error;
    }
  }

  // ========== Weekly Program Operations ==========

  /**
   * Get all weekly programs with their assigned days
   */
  async getAllPrograms(): Promise<WeeklyProgramWithDays[]> {
    try {
      return await WorkoutDao.getAllProgramsWithDays();
    } catch (error) {
      console.error('[WorkoutRepository] Failed to get all programs:', error);
      throw error;
    }
  }

  /**
   * Get the currently active program with its days
   */
  async getActiveProgram(): Promise<WeeklyProgramWithDays | null> {
    try {
      return await WorkoutDao.getActiveProgramWithDays();
    } catch (error) {
      console.error('[WorkoutRepository] Failed to get active program:', error);
      throw error;
    }
  }

  /**
   * Get a specific program by ID with its days
   */
  async getProgramById(programId: string): Promise<WeeklyProgramWithDays | null> {
    try {
      return await WorkoutDao.getProgramWithDaysById(programId);
    } catch (error) {
      console.error('[WorkoutRepository] Failed to get program by ID:', error);
      throw error;
    }
  }

  /**
   * Save a new weekly program or update existing one
   */
  async saveProgram(programWithDays: WeeklyProgramWithDays): Promise<void> {
    try {
      await WorkoutDao.insertProgramWithDays(programWithDays.program, programWithDays.days);
      console.log(`[WorkoutRepository] Saved weekly program: ${programWithDays.program.title}`);
    } catch (error) {
      console.error('[WorkoutRepository] Failed to save weekly program:', error);
      throw error;
    }
  }

  /**
   * Delete a weekly program
   */
  async deleteProgram(programId: string): Promise<void> {
    try {
      await WorkoutDao.deleteProgram(programId);
      console.log(`[WorkoutRepository] Deleted weekly program: ${programId}`);
    } catch (error) {
      console.error('[WorkoutRepository] Failed to delete weekly program:', error);
      throw error;
    }
  }

  /**
   * Activate a weekly program (deactivates all others)
   */
  async activateProgram(programId: string): Promise<void> {
    try {
      await WorkoutDao.activateProgram(programId);
      console.log(`[WorkoutRepository] Activated weekly program: ${programId}`);
    } catch (error) {
      console.error('[WorkoutRepository] Failed to activate weekly program:', error);
      throw error;
    }
  }

  // ========== Personal Record Operations ==========

  /**
   * Update personal record if the new performance is better
   */
  async updatePersonalRecordIfNeeded(
    exerciseId: string,
    weightPerCableKg: number,
    reps: number,
    workoutMode: string
  ): Promise<boolean> {
    try {
      const isNewPR = await PersonalRecordDao.updatePRIfBetter(
        exerciseId,
        weightPerCableKg,
        reps,
        workoutMode,
        Date.now()
      );

      if (isNewPR) {
        console.log(
          `[WorkoutRepository] New PR set for exercise ${exerciseId}: ${weightPerCableKg}kg x ${reps} reps (${workoutMode})`
        );
      }

      return isNewPR;
    } catch (error) {
      console.error('[WorkoutRepository] Failed to update personal record:', error);
      return false;
    }
  }

  // ========== Helper Methods ==========

  /**
   * Convert entity to WorkoutSession
   */
  private entityToWorkoutSession(entity: WorkoutSessionEntity): WorkoutSession {
    return {
      id: entity.id,
      timestamp: entity.timestamp,
      mode: entity.mode,
      reps: entity.reps,
      weightPerCableKg: entity.weightPerCableKg,
      progressionKg: entity.progressionKg,
      duration: entity.duration,
      totalReps: entity.totalReps,
      warmupReps: entity.warmupReps,
      workingReps: entity.workingReps,
      isJustLift: entity.isJustLift,
      stopAtTop: entity.stopAtTop,
      eccentricLoad: entity.eccentricLoad,
      echoLevel: entity.echoLevel,
      exerciseId: entity.exerciseId,
    };
  }

  /**
   * Convert entity to WorkoutMetric
   */
  private entityToWorkoutMetric(entity: WorkoutMetricEntity): WorkoutMetric {
    return {
      timestamp: entity.timestamp,
      loadA: entity.loadA,
      loadB: entity.loadB,
      positionA: entity.positionA,
      positionB: entity.positionB,
      ticks: entity.ticks,
    };
  }

  /**
   * Convert Routine to RoutineEntity
   */
  private routineToEntity(routine: Routine): RoutineEntity {
    return {
      id: routine.id,
      name: routine.name,
      description: routine.description,
      createdAt: routine.createdAt,
      lastUsed: routine.lastUsed,
      useCount: routine.useCount,
    };
  }

  /**
   * Convert RoutineExercise to RoutineExerciseEntity
   */
  private routineExerciseToEntity(exercise: RoutineExercise, routineId: string): RoutineExerciseEntity {
    // Determine mode string from workoutType
    let mode = 'OldSchool';
    let eccentricLoad = 100;
    let echoLevel = 1;

    if (exercise.workoutType.type === 'echo') {
      mode = 'Echo';
      eccentricLoad = exercise.workoutType.eccentricLoad;
      echoLevel = exercise.workoutType.level;
    } else if (exercise.workoutType.type === 'program') {
      mode = exercise.workoutType.mode.displayName.replace(/\s+/g, '');
    }

    return {
      id: exercise.id,
      routineId,
      exerciseName: exercise.exercise.name,
      exerciseMuscleGroup: exercise.exercise.muscleGroup,
      exerciseEquipment: exercise.exercise.equipment,
      exerciseDefaultCableConfig: exercise.exercise.defaultCableConfig,
      exerciseId: exercise.exercise.id,
      cableConfig: exercise.cableConfig,
      orderIndex: exercise.orderIndex,
      setReps: exercise.setReps.join(','),
      weightPerCableKg: exercise.weightPerCableKg,
      setWeights: exercise.setWeightsPerCableKg.join(','),
      mode,
      eccentricLoad,
      echoLevel,
      progressionKg: exercise.progressionKg,
      restSeconds: exercise.restSeconds,
      notes: exercise.notes,
      duration: exercise.duration,
    };
  }

  /**
   * Convert RoutineEntity to Routine
   */
  private entityToRoutine(entity: RoutineEntity, exerciseEntities: RoutineExerciseEntity[]): Routine {
    return {
      id: entity.id,
      name: entity.name,
      description: entity.description,
      exercises: exerciseEntities.map(this.entityToRoutineExercise),
      createdAt: entity.createdAt,
      lastUsed: entity.lastUsed,
      useCount: entity.useCount,
    };
  }

  /**
   * Convert RoutineExerciseEntity to RoutineExercise
   */
  private entityToRoutineExercise(entity: RoutineExerciseEntity): RoutineExercise {
    // Reconstruct Exercise
    const exercise: Exercise = {
      name: entity.exerciseName,
      muscleGroup: entity.exerciseMuscleGroup,
      equipment: entity.exerciseEquipment,
      defaultCableConfig: entity.exerciseDefaultCableConfig as CableConfiguration,
      id: entity.exerciseId,
    };

    // Parse setReps and setWeights
    const setReps = entity.setReps ? entity.setReps.split(',').map(Number) : [];
    const setWeights = entity.setWeights ? entity.setWeights.split(',').map(Number) : [];

    // Reconstruct WorkoutType
    let workoutType: WorkoutType;
    if (entity.mode === 'Echo') {
      workoutType = {
        type: 'echo',
        level: entity.echoLevel as EchoLevel,
        eccentricLoad: entity.eccentricLoad as EccentricLoad,
      };
    } else {
      // Map mode string to ProgramMode
      let programMode = ProgramMode.OldSchool;
      switch (entity.mode) {
        case 'Pump':
          programMode = ProgramMode.Pump;
          break;
        case 'TUT':
          programMode = ProgramMode.TUT;
          break;
        case 'TUTBeast':
          programMode = ProgramMode.TUTBeast;
          break;
        case 'EccentricOnly':
          programMode = ProgramMode.EccentricOnly;
          break;
        default:
          programMode = ProgramMode.OldSchool;
      }
      workoutType = { type: 'program', mode: programMode };
    }

    return {
      id: entity.id,
      exercise,
      cableConfig: entity.cableConfig as CableConfiguration,
      orderIndex: entity.orderIndex,
      setReps,
      weightPerCableKg: entity.weightPerCableKg,
      setWeightsPerCableKg: setWeights,
      workoutType,
      eccentricLoad: entity.eccentricLoad as EccentricLoad,
      echoLevel: entity.echoLevel as EchoLevel,
      progressionKg: entity.progressionKg,
      restSeconds: entity.restSeconds,
      notes: entity.notes,
      duration: entity.duration,
    };
  }

  /**
   * Generate UUID
   */
  private generateUUID(): string {
    return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c) => {
      const r = (Math.random() * 16) | 0;
      const v = c === 'x' ? r : (r & 0x3) | 0x8;
      return v.toString(16);
    });
  }
}

// Export singleton instance
let workoutRepositoryInstance: WorkoutRepositoryImpl | null = null;

export const getWorkoutRepository = (): IWorkoutRepository => {
  if (!workoutRepositoryInstance) {
    workoutRepositoryInstance = new WorkoutRepositoryImpl();
  }
  return workoutRepositoryInstance;
};

export const resetWorkoutRepository = (): void => {
  workoutRepositoryInstance = null;
};

// Export types
export type { IWorkoutRepository };
export { WorkoutRepositoryImpl };

File: deprecated/app/src/main/java/com/example/vitruvianredux/data/repository/WorkoutRepository.kt
package com.example.vitruvianredux.data.repository

import com.example.vitruvianredux.data.local.WorkoutDao
import com.example.vitruvianredux.data.local.WorkoutMetricEntity
import com.example.vitruvianredux.data.local.WorkoutSessionEntity
import com.example.vitruvianredux.data.local.RoutineEntity
import com.example.vitruvianredux.data.local.RoutineExerciseEntity
import com.example.vitruvianredux.data.local.WeeklyProgramWithDays
import com.example.vitruvianredux.data.local.PersonalRecordDao
import com.example.vitruvianredux.data.local.PersonalRecordEntity
import com.example.vitruvianredux.domain.model.WorkoutMetric
import com.example.vitruvianredux.domain.model.WorkoutSession
import com.example.vitruvianredux.domain.model.Routine
import com.example.vitruvianredux.domain.model.RoutineExercise
import com.example.vitruvianredux.domain.model.Exercise
import com.example.vitruvianredux.domain.model.CableConfiguration
import com.example.vitruvianredux.domain.model.WorkoutMode
import com.example.vitruvianredux.domain.model.WorkoutType
import com.example.vitruvianredux.domain.model.ProgramMode
import com.example.vitruvianredux.domain.model.EchoLevel
import com.example.vitruvianredux.domain.model.EccentricLoad
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.firstOrNull
import kotlinx.coroutines.flow.map
import timber.log.Timber
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Repository for workout history management
 */
@Singleton
class WorkoutRepository @Inject constructor(
    private val workoutDao: WorkoutDao,
    private val personalRecordDao: PersonalRecordDao
) {
    
    /**
     * Save a workout session
     */
    suspend fun saveSession(session: WorkoutSession): Result<Unit> {
        return try {
            val entity = WorkoutSessionEntity(
                id = session.id,
                timestamp = session.timestamp,
                mode = session.mode,
                reps = session.reps,
                weightPerCableKg = session.weightPerCableKg,
                progressionKg = session.progressionKg,
                duration = session.duration,
                totalReps = session.totalReps,
                warmupReps = session.warmupReps,
                workingReps = session.workingReps,
                isJustLift = session.isJustLift,
                stopAtTop = session.stopAtTop,
                eccentricLoad = session.eccentricLoad,
                echoLevel = session.echoLevel,
                exerciseId = session.exerciseId
            )
            workoutDao.insertSession(entity)
            Timber.d("Saved workout session: ${session.id}")
            Result.success(Unit)
        } catch (e: Exception) {
            Timber.e(e, "Failed to save workout session")
            Result.failure(e)
        }
    }
    
    /**
     * Save workout metrics (batch insert for performance)
     */
    suspend fun saveMetrics(sessionId: String, metrics: List<WorkoutMetric>): Result<Unit> {
        return try {
            val entities = metrics.map { metric ->
                WorkoutMetricEntity(
                    sessionId = sessionId,
                    timestamp = metric.timestamp,
                    loadA = metric.loadA,
                    loadB = metric.loadB,
                    positionA = metric.positionA,
                    positionB = metric.positionB,
                    ticks = metric.ticks
                )
            }
            workoutDao.insertMetrics(entities)
            Timber.d("Saved ${entities.size} metrics for session $sessionId")
            Result.success(Unit)
        } catch (e: Exception) {
            Timber.e(e, "Failed to save workout metrics")
            Result.failure(e)
        }
    }
    
    /**
     * Get all workout sessions
     */
    fun getAllSessions(): Flow<List<WorkoutSession>> {
        return workoutDao.getAllSessions().map { entities ->
            entities.map { it.toWorkoutSession() }
        }
    }
    
    /**
     * Get recent workout sessions
     */
    fun getRecentSessions(limit: Int = 10): Flow<List<WorkoutSession>> {
        return workoutDao.getRecentSessions(limit).map { entities ->
            entities.map { it.toWorkoutSession() }
        }
    }
    
    /**
     * Get a specific workout session
     */
    suspend fun getSession(sessionId: String): WorkoutSession? {
        return workoutDao.getSession(sessionId)?.toWorkoutSession()
    }
    
    /**
     * Get metrics for a workout session
     */
    fun getMetricsForSession(sessionId: String): Flow<List<WorkoutMetric>> {
        return workoutDao.getMetricsForSession(sessionId).map { entities ->
            entities.map { it.toWorkoutMetric() }
        }
    }
    
    /**
     * Delete a workout
     */
    suspend fun deleteWorkout(sessionId: String): Result<Unit> {
        return try {
            workoutDao.deleteWorkout(sessionId)
            Timber.d("Deleted workout: $sessionId")
            Result.success(Unit)
        } catch (e: Exception) {
            Timber.e(e, "Failed to delete workout")
            Result.failure(e)
        }
    }
    
    /**
     * Delete all workouts
     */
    suspend fun deleteAllWorkouts(): Result<Unit> {
        return try {
            workoutDao.deleteAllWorkouts()
            Timber.d("Deleted all workouts")
            Result.success(Unit)
        } catch (e: Exception) {
            Timber.e(e, "Failed to delete all workouts")
            Result.failure(e)
        }
    }
    
    // ========== Routine Operations ==========
    
    /**
     * Save a routine with exercises
     */
    suspend fun saveRoutine(routine: Routine): Result<Unit> {
        return try {
            val entity = routine.toEntity()
            val exerciseEntities = routine.exercises.map { it.toEntity(routine.id) }
            workoutDao.insertRoutineWithExercises(entity, exerciseEntities)
            Timber.d("Saved routine: ${routine.name}")
            Result.success(Unit)
        } catch (e: Exception) {
            Timber.e(e, "Failed to save routine")
            Result.failure(e)
        }
    }
    
    /**
     * Update a routine
     */
    suspend fun updateRoutine(routine: Routine): Result<Unit> {
        return try {
            val entity = routine.toEntity()
            val exerciseEntities = routine.exercises.map { it.toEntity(routine.id) }
            workoutDao.updateRoutineWithExercises(entity, exerciseEntities)
            Timber.d("Updated routine: ${routine.name}")
            Result.success(Unit)
        } catch (e: Exception) {
            Timber.e(e, "Failed to update routine")
            Result.failure(e)
        }
    }
    
    /**
     * Get all routines
     */
    fun getAllRoutines(): Flow<List<Routine>> {
        return workoutDao.getAllRoutines().map { entities ->
            entities.map { entity ->
                val exercises = workoutDao.getExercisesForRoutineSync(entity.id)
                entity.toRoutine(exercises)
            }
        }
    }
    
    /**
     * Get a specific routine
     */
    suspend fun getRoutine(routineId: String): Routine? {
        return try {
            val entity = workoutDao.getRoutineById(routineId) ?: return null
            val exercises = workoutDao.getExercisesForRoutineSync(routineId)
            entity.toRoutine(exercises)
        } catch (e: Exception) {
            Timber.e(e, "Failed to get routine")
            null
        }
    }
    
    /**
     * Delete a routine
     */
    suspend fun deleteRoutine(routineId: String): Result<Unit> {
        return try {
            workoutDao.deleteRoutineComplete(routineId)
            Timber.d("Deleted routine: $routineId")
            Result.success(Unit)
        } catch (e: Exception) {
            Timber.e(e, "Failed to delete routine")
            Result.failure(e)
        }
    }
    
    /**
     * Mark routine as used (updates lastUsed and increments useCount)
     */
    suspend fun markRoutineUsed(routineId: String): Result<Unit> {
        return try {
            workoutDao.markRoutineUsed(routineId)
            Timber.d("Marked routine used: $routineId")
            Result.success(Unit)
        } catch (e: Exception) {
            Timber.e(e, "Failed to mark routine used")
            Result.failure(e)
        }
    }

    /**
     * Get routine by ID as a Flow
     */
    fun getRoutineById(routineId: String): Flow<Routine?> {
        return workoutDao.observeRoutineById(routineId).map { entity ->
            entity?.let {
                val exercises = workoutDao.getExercisesForRoutineSync(routineId)
                it.toRoutine(exercises)
            }
        }
    }

    // ========== Weekly Programs ==========

    /**
     * Get all weekly programs with their assigned days
     */
    fun getAllPrograms(): Flow<List<WeeklyProgramWithDays>> =
        workoutDao.getAllProgramsWithDays()

    /**
     * Get the currently active program with its days
     */
    fun getActiveProgram(): Flow<WeeklyProgramWithDays?> =
        workoutDao.getActiveProgramWithDays()

    /**
     * Get a specific program by ID with its days
     */
    fun getProgramById(programId: String): Flow<WeeklyProgramWithDays?> =
        workoutDao.getProgramWithDaysById(programId)

    /**
     * Save a new weekly program or update existing one
     */
    suspend fun saveProgram(programWithDays: WeeklyProgramWithDays): Result<Unit> {
        return try {
            workoutDao.insertProgramWithDays(
                program = programWithDays.program,
                days = programWithDays.days
            )
            Timber.d("Saved weekly program: ${programWithDays.program.title}")
            Result.success(Unit)
        } catch (e: Exception) {
            Timber.e(e, "Failed to save weekly program")
            Result.failure(e)
        }
    }

    /**
     * Delete a weekly program
     */
    suspend fun deleteProgram(programId: String): Result<Unit> {
        return try {
            workoutDao.deleteProgram(programId)
            Timber.d("Deleted weekly program: $programId")
            Result.success(Unit)
        } catch (e: Exception) {
            Timber.e(e, "Failed to delete weekly program")
            Result.failure(e)
        }
    }

    /**
     * Activate a weekly program (deactivates all others)
     */
    suspend fun activateProgram(programId: String): Result<Unit> {
        return try {
            workoutDao.activateProgram(programId)
            Timber.d("Activated weekly program: $programId")
            Result.success(Unit)
        } catch (e: Exception) {
            Timber.e(e, "Failed to activate weekly program")
            Result.failure(e)
        }
    }

    // ========== Personal Records ==========

    /**
     * Get all personal records
     */
    fun getAllPersonalRecords(): Flow<List<PersonalRecordEntity>> =
        personalRecordDao.getAllPRs()

    /**
     * Update personal record if the new performance is better
     */
    suspend fun updatePersonalRecordIfNeeded(
        exerciseId: String,
        weightPerCableKg: Float,
        reps: Int,
        workoutMode: String
    ): Boolean {
        return try {
            val isNewPR = personalRecordDao.updatePRIfBetter(
                exerciseId = exerciseId,
                weightPerCableKg = weightPerCableKg,
                reps = reps,
                workoutMode = workoutMode,
                timestamp = System.currentTimeMillis()
            )
            if (isNewPR) {
                Timber.d("New PR set for exercise $exerciseId: ${weightPerCableKg}kg x $reps reps ($workoutMode)")
            }
            isNewPR
        } catch (e: Exception) {
            Timber.e(e, "Failed to update personal record")
            false
        }
    }
}

// Extension functions for mapping between entities and domain models
private fun WorkoutSessionEntity.toWorkoutSession() = WorkoutSession(
    id = id,
    timestamp = timestamp,
    mode = mode,
    reps = reps,
    weightPerCableKg = weightPerCableKg,
    progressionKg = progressionKg,
    duration = duration,
    totalReps = totalReps,
    warmupReps = warmupReps,
    workingReps = workingReps,
    isJustLift = isJustLift,
    stopAtTop = stopAtTop,
    eccentricLoad = eccentricLoad,
    echoLevel = echoLevel,
    exerciseId = exerciseId
)

private fun WorkoutMetricEntity.toWorkoutMetric() = WorkoutMetric(
    timestamp = timestamp,
    loadA = loadA,
    loadB = loadB,
    positionA = positionA,
    positionB = positionB,
    ticks = ticks
)

// Routine mapping extensions
private fun Routine.toEntity() = RoutineEntity(
    id = id,
    name = name,
    description = description,
    createdAt = createdAt,
    lastUsed = lastUsed,
    useCount = useCount
)

private fun RoutineExercise.toEntity(routineId: String) = RoutineExerciseEntity(
    id = id,
    routineId = routineId,
    // Store Exercise data class fields
    exerciseName = exercise.name,
    exerciseMuscleGroup = exercise.muscleGroup,
    exerciseEquipment = exercise.equipment,
    exerciseDefaultCableConfig = exercise.defaultCableConfig.name, // Convert enum to String
    exerciseId = exercise.id, // Store exercise library ID
    // Routine-specific configuration
    cableConfig = cableConfig.name, // Convert enum to String
    orderIndex = orderIndex,
    setReps = setReps.joinToString(","), // Convert List<Int> to comma-separated String
    weightPerCableKg = weightPerCableKg,
    setWeights = setWeightsPerCableKg.joinToString(",") { it.toString() },
    mode = when (workoutType) {
        is WorkoutType.Program -> when (workoutType.mode) {
            is ProgramMode.OldSchool -> "OldSchool"
            is ProgramMode.Pump -> "Pump"
            is ProgramMode.TUT -> "TUT"
            is ProgramMode.TUTBeast -> "TUTBeast"
            is ProgramMode.EccentricOnly -> "EccentricOnly"
        }
        is WorkoutType.Echo -> "Echo"
    },
    eccentricLoad = when (workoutType) {
        is WorkoutType.Echo -> workoutType.eccentricLoad.percentage
        else -> eccentricLoad.percentage
    },
    echoLevel = when (workoutType) {
        is WorkoutType.Echo -> workoutType.level.levelValue
        else -> echoLevel.levelValue
    },
    progressionKg = progressionKg,
    restSeconds = restSeconds,
    notes = notes,
    duration = duration
)

private fun RoutineEntity.toRoutine(exerciseEntities: List<RoutineExerciseEntity>) = Routine(
    id = id,
    name = name,
    description = description,
    exercises = exerciseEntities.map { it.toRoutineExercise() },
    createdAt = createdAt,
    lastUsed = lastUsed,
    useCount = useCount
)

private fun RoutineExerciseEntity.toRoutineExercise() = RoutineExercise(
    id = id,
    // Reconstruct Exercise data class from stored fields
    exercise = Exercise(
        name = exerciseName,
        muscleGroup = exerciseMuscleGroup,
        equipment = exerciseEquipment,
        defaultCableConfig = CableConfiguration.valueOf(exerciseDefaultCableConfig),
        id = exerciseId  // Pass through exercise library ID
    ),
    cableConfig = CableConfiguration.valueOf(cableConfig), // Convert String to enum
    orderIndex = orderIndex,
    setReps = if (setReps.isEmpty()) emptyList() else setReps.split(",").mapNotNull { it.toIntOrNull() },
    weightPerCableKg = weightPerCableKg,
    setWeightsPerCableKg = if (setWeights.isEmpty()) emptyList() else setWeights.split(",").mapNotNull { it.toFloatOrNull() },
    workoutType = when (mode) {
        "Echo" -> WorkoutType.Echo(
            level = EchoLevel.values().getOrNull(echoLevel) ?: EchoLevel.HARDER,
            eccentricLoad = EccentricLoad.values().find { it.percentage == eccentricLoad } ?: EccentricLoad.LOAD_100
        )
        "OldSchool" -> WorkoutType.Program(ProgramMode.OldSchool)
        "Pump" -> WorkoutType.Program(ProgramMode.Pump)
        "TUT" -> WorkoutType.Program(ProgramMode.TUT)
        "TUTBeast" -> WorkoutType.Program(ProgramMode.TUTBeast)
        "EccentricOnly" -> WorkoutType.Program(ProgramMode.EccentricOnly)
        else -> WorkoutType.Program(ProgramMode.OldSchool)
    },
    eccentricLoad = EccentricLoad.values().find { it.percentage == eccentricLoad } ?: EccentricLoad.LOAD_100,
    echoLevel = EchoLevel.values().getOrNull(echoLevel) ?: EchoLevel.HARDER,
    progressionKg = progressionKg,
    restSeconds = restSeconds,
    notes = notes,
    duration = duration
)
